pkg/dns/dns_test.go

```diff
@@ -0,0 +1,377 @@
+
+	// add endpoints
+	kd.handleEndpointAdd(endpoints)
+
+	assertDNSForHeadlessService(t, kd, endpoints)
+
+	// remove service
+	kd.removeService(service)
+	assertNoDNSForHeadlessService(t, kd, service)
+}
+
+func newService(namespace, serviceName, clusterIP, portName string, portNumber int32) *kapi.Service {
+	service := kapi.Service{
+		ObjectMeta: kapi.ObjectMeta{
+			Name:      serviceName,
+			Namespace: namespace,
+		},
+		Spec: kapi.ServiceSpec{
+			ClusterIP: clusterIP,
+			Ports: []kapi.ServicePort{
+				{Port: portNumber, Name: portName, Protocol: "TCP"},
+			},
+		},
+	}
+	return &service
+}
+
+func newHeadlessService() *kapi.Service {
+	service := kapi.Service{
+		ObjectMeta: kapi.ObjectMeta{
+			Name:      testService,
+			Namespace: testNamespace,
+		},
+		Spec: kapi.ServiceSpec{
+			ClusterIP: "None",
+			Ports: []kapi.ServicePort{
+				{Port: 0},
+			},
+		},
+	}
+	return &service
+}
+
+func newEndpoints(service *kapi.Service, subsets ...kapi.EndpointSubset) *kapi.Endpoints {
+	endpoints := kapi.Endpoints{
+		ObjectMeta: service.ObjectMeta,
+		Subsets:    []kapi.EndpointSubset{},
+	}
+
+	for _, subset := range subsets {
+		endpoints.Subsets = append(endpoints.Subsets, subset)
+	}
+	return &endpoints
+}
+
+func newSubsetWithOnePort(portName string, port int32, ips ...string) kapi.EndpointSubset {
+	subset := newSubset()
+	subset.Ports = append(subset.Ports, kapi.EndpointPort{Port: port, Name: portName, Protocol: "TCP"})
+	for _, ip := range ips {
+		subset.Addresses = append(subset.Addresses, kapi.EndpointAddress{IP: ip})
+	}
+	return subset
+}
+
+func newSubsetWithTwoPorts(portName1 string, portNumber1 int32, portName2 string, portNumber2 int32, ips ...string) kapi.EndpointSubset {
+	subset := newSubsetWithOnePort(portName1, portNumber1, ips...)
+	subset.Ports = append(subset.Ports, kapi.EndpointPort{Port: portNumber2, Name: portName2, Protocol: "TCP"})
+	return subset
+}
+
+func newSubset() kapi.EndpointSubset {
+	subset := kapi.EndpointSubset{
+		Addresses: []kapi.EndpointAddress{},
+		Ports:     []kapi.EndpointPort{},
+	}
+	return subset
+}
+
+func assertSRVForHeadlessService(t *testing.T, kd *KubeDNS, s *kapi.Service, e *kapi.Endpoints) {
+	for _, subset := range e.Subsets {
+		for _, port := range subset.Ports {
+			records, err := kd.Records(getSRVFQDN(kd, s, port.Name), false)
+			require.NoError(t, err)
+			assertRecordPortsMatchPort(t, port.Port, records)
+			assertCNameRecordsMatchEndpointIPs(t, kd, subset.Addresses, records)
+		}
+	}
+}
+
+func assertDNSForHeadlessService(t *testing.T, kd *KubeDNS, e *kapi.Endpoints) {
+	records, err := kd.Records(getEndpointsFQDN(kd, e), false)
+	require.NoError(t, err)
+	endpoints := map[string]bool{}
+	for _, subset := range e.Subsets {
+		for _, endpointAddress := range subset.Addresses {
+			endpoints[endpointAddress.IP] = true
+		}
+	}
+	assert.Equal(t, len(endpoints), len(records))
+	for _, record := range records {
+		_, found := endpoints[record.Host]
```

Comment:
I believe we do it already, just not in the assert.. function since the function should not be mutating anything..